### PR TITLE
Rework ReactiveStore to prevent crashes

### PR DESCRIPTION
**Pull request by szabinst**
Rework ReactiveStore to prevent crashes

refs: MBL-17259
affects: Student, Parent, Teacher
release note: Fixed an issue that caused occasional crashes.
test plan: Test the whole app if everything works as before.

## Checklist

- [ ] Tested on phone
- [ ] Tested on tablet

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Parent Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>



#### Release Note: 
Fixed an issue that caused occasional crashes.

#### Affected Apps: Student, Parent, Teacher

[MBL-17259](https://instructure.atlassian.net/browse/MBL-17259)

<details><summary>:x: XCTest failed: CoreTests/CourseSyncAnnouncementsInteractorLiveTests/testFailuresReported</summary>


    failed - Received unexpected output from publisher.
    failed - Unexpected finish event while waiting on failure.
    failed - Received unexpected output from publisher.
    failed - Unexpected finish event while waiting on failure.
    failed - Received unexpected output from publisher.
    failed - Unexpected finish event while waiting on failure.
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8cdb5b8338322587e31191bdccc9d1e614a2c998
</p>

**Review by vargaat — CHANGES_REQUESTED** [review]
Good job on making the store stateless, this is much safer to use now. I had some comments on the parameter naming and the function doc wording, let me know your thoughts.

File: Core/Core/Store/ReactiveStore.swift [hunk 1]
```diff
@@ -75,166 +36,145 @@ public class ReactiveStore<U: UseCase> {
...
+    /// - Parameters:
+    ///     - forceFetch: Tells the fetch request if it should get data from the API or from cache. Defaults to **false**.
+    ///     - loadAllPages: Tells the fetch request if it should load all the pages or just the first one. Defaults to **true**.
+    ///     - publishDatabaseChanges: Tells the fetch request to keep observing database changes after the API response is downloaded and saved. Defaults to **false**.
+    ///
+    /// - Returns: A list of entities or an error.
+    public func getEntities(
+        forceFetch: Bool = false,
+        loadAllPages: Bool = true,
+        publishDatabaseChanges: Bool = false
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Store/ReactiveStore.swift:54]: I actually named this property exactly `keepObservingDatabaseChanges` at first, but changed it later on. I'll change it back.

File: Core/Core/Store/ReactiveStore.swift [hunk 2]
```diff
@@ -75,166 +36,145 @@ public class ReactiveStore<U: UseCase> {
...
-    public func observeEntities(forceFetch: Bool = false, loadAllPages: Bool = false) -> AnyPublisher<State, Never> {
-        cancellable?.cancel()
-        cancellable = nil
-
+    /// Produces a list of entities for the given UseCase.
+    /// When the device is connected to the internet, it will fetch data from the API by downloading all pages, unless specificied differently. The result is then cached to the database
+    /// emitted by the publisher.
+    /// When the device is offline, it will read data from Core Data.
+    /// - Parameters:
+    ///     - forceFetch: Tells the fetch request if it should get data from the API or from cache. Defaults to **false**.
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Store/ReactiveStore.swift]: Updated the parameter name.

File: Core/Core/Store/ReactiveStore.swift [hunk 3]
```diff
@@ -75,166 +36,145 @@ public class ReactiveStore<U: UseCase> {
...
-    }
-
-    /// Calling this function will keep the instance in memory until `cancel()` is called. The recommended approach is calling `cancel()` from the interactor's deinit function.
-    public func observeEntities(forceFetch: Bool = false, loadAllPages: Bool = false) -> AnyPublisher<State, Never> {
-        cancellable?.cancel()
-        cancellable = nil
-
+    /// Produces a list of entities for the given UseCase.
+    /// When the device is connected to the internet, it will fetch data from the API by downloading all pages, unless specificied differently. The result is then cached to the database
+    /// emitted by the publisher.
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Store/ReactiveStore.swift]: Reworded the description.

**Review by vargaat — APPROVED** [review]
QA+1

**Pull request merged**